src/vsi.rs

```diff
@@ -5,7 +5,29 @@ use std::path::{Path, PathBuf};
 use gdal_sys::{VSIFCloseL, VSIFileFromMemBuffer, VSIFree, VSIGetMemFileBuffer, VSIUnlink};
 
 use crate::errors::{GdalError, Result};
-use crate::utils::{_last_null_pointer_err, _path_to_c_string};
+use crate::utils::{_last_null_pointer_err, _path_to_c_string, _string_array};
+
+/// Read the file names from a virtual file system with optional recursion.
+pub fn read_dir<P: AsRef<Path>>(path: P, recursive: bool) -> Result<Vec<String>> {
+    let path = _path_to_c_string(path.as_ref())?;
+    let data = unsafe {
+        if recursive {
+            let data = gdal_sys::VSIReadDirRecursive(path.as_ptr());
```

Comment:
Nit:

```suggestion
            let data = unsafe { gdal_sys::VSIReadDirRecursive(path.as_ptr()) }; 
```

and the same below.

Smaller `unsafe` blocks are nicer because they make it more obvious what exactly is unsafe.